lighthouse-core/report/html/renderer/code-snippet-renderer.js

```diff
@@ -0,0 +1,260 @@
+   */
+  static renderOmittedLines(dom, tmpl, classOptions = {}) {
+    return CodeSnippetRenderer.renderLine(dom, tmpl, {
+      number: '…',
+      content: '',
+    }, classOptions);
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippet} details
+   * @return {DocumentFragment}
+   */
+  static renderSnippet(dom, tmpl, details) {
+    const {highlights, lineCount, lines} = details;
+
+    const template = dom.cloneTemplate('#tmpl-lh-code-snippet__content', tmpl);
+    const snippet = dom.find('.lh-code-snippet__snippet-inner', template);
+
+    const nonLineSpecificHighlights = highlights.filter(h => typeof h.lineNumber !== 'number');
+    nonLineSpecificHighlights.forEach(h =>
+      snippet.append(CodeSnippetRenderer.renderHighlightMessage(dom, tmpl, h))
+    );
+
+    snippet.append(CodeSnippetRenderer.renderSnippetLines(dom, tmpl, details));
+
+    // If expanded view still doesn't include all lines then indicate that that
+    const firstLineIsVisible = lines[0].number === 1;
+    const lastLineIsVisible = lines.slice(-1)[0].number === lineCount;
+    if (!firstLineIsVisible) {
+      snippet.prepend(CodeSnippetRenderer.renderOmittedLines(dom, tmpl, {expandedOnly: true}));
+    }
+    if (!lastLineIsVisible) {
+      snippet.append(CodeSnippetRenderer.renderOmittedLines(dom, tmpl, {expandedOnly: true}));
+    }
+
+    return template;
+  }
+
+  /**
+   * @param {DOM} dom
+   * @param {DocumentFragment} tmpl
+   * @param {LH.Audit.DetailsRendererCodeSnippet} details
+   * @returns {DocumentFragment}
+   */
+  static renderSnippetLines(dom, tmpl, details) {
+    const {highlights, lineCount, lines} = details;
+    const collapsedLines = Util.filterRelevantLines(lines, highlights, 2);
+    const hasLineSpecificHighlights = highlights.some(h => typeof h.lineNumber === 'number');
+    const hasOnlyNonLineSpecficHighlights = highlights.length > 0 && !hasLineSpecificHighlights;
+
+    const lineContainer = dom.createFragment();
+
+    // Keep track if a highlighted line has been seen, because in collapsed mode
+    // we don't show the omitted lines indication before the first visible set of lines
+    let hasSeenHighlight = false;
+    for (let lineNumber = 1; lineNumber <= lineCount; lineNumber++) {
+      const {line, previousLine} = getLineAndPreviousLine(lines, lineNumber);
+      const {line: collapsedLine, previousLine: collapsedPreviousLine} =
+        getLineAndPreviousLine(collapsedLines, lineNumber);
+
+      if (hasSeenHighlight) {
+        // Show if some lines were omitted
+        if (line && !previousLine) {
+          lineContainer.append(
+            CodeSnippetRenderer.renderOmittedLines(dom, tmpl, {expandedOnly: true}));
+        }
+        if (collapsedLine && !collapsedPreviousLine) {
+          lineContainer.append(
+            CodeSnippetRenderer.renderOmittedLines(dom, tmpl, {collapsedOnly: true}));
+        }
+      }
+
+      if (!line) {
```

Comment:
is this when we iterate through the omitted lines?

Reply:
Yeah, we skip the lines where we don't know the content.